### ISSUE TITLE
[REQUEST]: Prefill compiler options edit box

### ISSUE THREAD

**Issue by ojeda** [compiler-explorer/compiler-explorer#3100]
### Is your feature request related to a problem? Please describe

Currently, the compiler options edit box is empty, but there are some flags that may be useful to prefill. For instance:

  - Some compilers like `rustc` default to the oldest edition of the Rust language for backwards compatibility, while I would presume most folks would like to have access to the latest by default.

  - Other compilers use a language version default that changes over time (i.e. with different compiler versions) and also enables extensions by default, e.g. GCC for C. In particular, since GCC is the default C and C++ compiler, it may be best to be explicit and compile the language in its standard form.

  - There are other flags like warnings or optimizations that may be useful for the majority of users.

If a user needs other flags, then it is arguably faster to change the prefilled ones than type them from scratch (plus it serves as a quick reminder for those not using a particular toolchain every day and/or newcomers to a language).

### Describe the solution you'd like

Prefill the compiler options edit box with a predefined string per language set in the config, e.g.:

```properties
default_user_options=-std=c++20 -Wall -Wextra -O2
```

Groups of options that may be worth considering:
  - Standard / edition / language flags.
  - Warning / diagnostic flags.
  - Optimization flags.

For instance:
  - C (i.e. GCC): `-std=c17 -Wall -Wextra -O2`
  - C++ (i.e. GCC): `-std=c++20 -Wall -Wextra -O2`
  - Rust (i.e. rustc): `--edition=2021 -O`

However, a problem is that currently we do not change the options when switching to a different compiler (only when switching language they are reset). Thus it may be useful to have a set of options also per compiler group and/or version, for instance:
  - C (GCC, Clang, icc, icx...): `-std=c17 -Wall -Wextra -O2`
  - C++ (GCC, Clang, icc, icx...): `-std=c++20 -Wall -Wextra -O2`
  - C (MSVC): `/std:c17 /permissive- /W4 /O2`
  - C++ (MSVC): `/std:c++20 /permissive- /W4 /O2`
  - Rust (rustc): `--edition=2021 -O`

Done with something like:

```properties
# ...
group.gcc86.default_user_options=-Wall -Wextra -O2
# Particular
compiler.g94.default_user_options=-std=c++17
# ...
compiler.g112.default_user_options=-std=c++20
```

This would allow us to give a user useful flags for every toolchain when switching (or simply adding) compilers. The question is when/how to switch the flags. Several alternatives:
  - If switching to an incompatible compiler, use the prefill; otherwise, no (e.g. from GCC to MSVC).
  - Only switch if the edit box contents match the prefill.
  - Only switch if the edit box was not modified by the user.

This could also be useful to replace some existing usage of `.options=` in the config, perhaps.

### Describe alternatives you've considered

Not applicable.

### Additional context

Not applicable.

**Comment by rmccampbell** [comment]
It seems that #5001 addresses setting the C++ std version for Clang and GCC, but it doesn't work for MSVC. I would love it if there was a way I could set default std versions for all (major) compilers and not have to change the incompatible flags every time I change compilers.

**Comment by ojeda** [comment]
Yeah, I think letting users edit these defaults would be nice (even if the defaults are the empty set in many cases). There is an ongoing discussion about flags, overrides and pre-filling where I mentioned that a couple weeks ago (https://github.com/compiler-explorer/compiler-explorer/issues/3765#issuecomment-1668478130).

I have edited OP to include the user-editable parts and simplified the wording of the proposal.
